spec/21-http_header_format_rationale.md

```diff
@@ -16,6 +16,103 @@ Making `trace-flags` optional doesn't save a lot, but makes specification more c
 
 We were using the term `span-id` in the `traceparent`, but not all tracing systems are built around span model, e.g. X-Trace, Canopy, SolarWinds, are built around event model, which is considered more expressive than the span model. There is nothing in the spec actually requires the model to be span-based, and passing the ID of the happened-before "thing" should work for both types of trace models. We considered names `call-id`, `request-id`. However out of all replacements `parent-id` is probably the best name. First, it matched the header name. Second it indicates a difference between caller and callee. Discussing AMQP we realized that `message-id` header defined by AMQP refers to individual message, and semantically not the same as traceparent. Message id can be used to dedup messages on the server when traceparent only defines the source this message came from.
 
+## Trace ID size
+
+In high load apps 64 bits is not enough to guarantee enough uniqueness of
+`trace-id` over a typical period of time - say 72 hours. That's said 128 bit
+`trace-id` may provide an excessive randomness for a smaller apps. However, in
+modern world many apps are using cloud services and SaaS components that may be
+shared by numerous smaller apps. So if those apps are using smaller `trace-id`,
+cloud services may not be able to correlate incoming requests from those apps to
+the proper distributed trace inside the cloud component.
+
+Thus for improved interoperability, this specification defines `trace-id` as a
+128 bit array of bytes.
+
+## Trace ID and interoperability with 64bit systems
+
+There are systems today using 64-bit `trace-id`s. These systems are not always
+easy to switch to longer `trace-id` to confirm to this specification
+requirement. The cost of this switch can be prohibitively expensive from both -
+backend capacity and indexes as well as in-process propagation limits.
+
+When addressing interoperability with these systems requirements the following
+is taken into consideration:
+
+1. The main objective of the specification is to promote interoperability of
```

Comment:
If these points are true, then shouldn't the recommendation be to backfill with zeros, not another random 64 bits? Wouldn't that promote better interoperability with brown field ecosystems that are currently on 64 bit, with the way forward being that those systems would organically move to 128 bit as they can? Currently, by recommending backfilling with random 64 bits, the spec is encouraging breakage (and it's just surprising/confusing/unexpected to see backfilling with randomness - I would expect something deterministic, and the brown field has chosen the deterministic path of backfilling with zeros).

Reply:
further in the doc there is a note that backfilling with random numbers will encourage you to test that those numbers will be propagated. And 64 system will not break the 128 bit system passing trace thru it. Spec is encouraging interoperability of different systems, which don't know about each other.

Reply:
> encourage you to test

This doesn't seem like a good enough reason to break compatibility with some 64-bit systems. The spec can encourage you to test by stating so in the spec. I'm not sure I can see many spec implementors who would only test because the random numbers encouraged testing, and would fail to test if it was simply stated in the spec.